### ISSUE TITLE
Configure Analyze Regions

### ISSUE THREAD

**Issue by tischi** [ijpb/MorphoLibJ#60]
@dlegland 

How does one set [those booleans](https://github.com/ijpb/MorphoLibJ/blob/59b691fb8fa457e2fad8d5eb57c82c0fe6115e67/src/main/java/inra/ijpb/plugins/AnalyzeRegions.java#L59) from Java?

I mean [without calling the UI](https://github.com/ijpb/MorphoLibJ/blob/59b691fb8fa457e2fad8d5eb57c82c0fe6115e67/src/main/java/inra/ijpb/plugins/AnalyzeRegions.java#L114), but programmatically....

**Comment by dlegland** [comment]
Hi @tischi,
I do not think this is possible. When using an ImageJ plugin, one has to rely on the "run" method.
One way could be to split the implementation, such that the plugin class only manages the GUI. I plan to refactor the Analyze regions plugins in a near future, so maybe I can also add this feature, if it is possible and not too hard.

**Comment by tischi** [comment]
Thanks! A very easy fix would be to make the boolean variables `public`. What do you think?

**Comment by dlegland** [comment]
Hi,
yes, you're right, this could be a quick fix! I try to have a look soon!

**Comment by tischi** [comment]
@dlegland any progress on this front? :-)

**Comment by dlegland** [comment]
Hi,
not much at the moment, 
but I plan to check this when I come back from holidays. hopefully end of august or beginning of september...


best,
David

**Comment by dlegland** [comment]
Hi,
I just pushed [a commit](https://github.com/ijpb/MorphoLibJ/commit/43c3891f6fa3c5b584e15c8b3081843321977059) that updates the AnayzeRegions plugins. 

It introduces an inner class "Features", that contains public boolean fields for selecting the features to compute. Example:
```
// creates a new Features instance to select the features to compute.  
AnalyzeRegions.Features features = new AnalyzeRegions.Features();
features.setAll(false);
features.area = true;
features.perimeter = true;
features.centroid = true;
    
// compute the features, and returns the corresponding table
ResultsTable table = AnalyzeRegions.process(imagePlus, features);
tableshow("Morphometry");
```

Feedback welcome!

**Comment by dlegland** [comment]
Proposition from @oburri: use a Builder type approach.

    ResultsTable table = AnalyzeRegions.create( imagePlus )
           .getArea()
           .getPerimeter()
           .getCentroid()
           .compute();

**Comment by lacan** [comment]
As a suggestion for this refactoring, I would go for a Builder type approach…
```
ResultsTable table = AnalyzeRegions.create( imagePlus )
        .getArea()
        .getPerimeter()
        .getCentroid()
        .compute()
```
Or something along those lines :-) A bit like the [ResultsBuilder](https://github.com/ijpb/MorphoLibJ/blob/8d31dada3d186bcdbacd845f41aa3e317c79e381/src/main/java/inra/ijpb/measure/ResultsBuilder.java)

**Comment by tischi** [comment]
@dlegland 

0. Thank you very much! Looks beautiful!
1. What happens if one does this:  `ResultsTable table = new MorphometricFeatures2D().computeTable(imagePlus);` does one get **all** or **none**?
2. Will that also work in 3D with the same API?
3. _finally used a collection features (instances of the inner "Feature" enum)._ I am intrigued, could you post a link to the code?

**Comment by dlegland** [comment]
> Thank you very much! Looks beautiful!

Thank you! I hope this will be more scriptable than the AnalyzeRegions plugin, and that this will not be too complicated to maintain by introducing new features...

> What happens if one does this: ResultsTable table = new MorphometricFeatures2D().computeTable(imagePlus); does one get all or none?

One get an empty table. Only the region label is provided. 

> Will that also work in 3D with the same API?

Yes! working on it... nearly ready, I just would like to enhance unit tests and general checks.

> finally used a collection features (instances of the inner "Feature" enum). I am intrigued, could you post a link to the code?

Sure! the main class is [MorphometricFeatures2D.java](https://github.com/ijpb/MorphoLibJ/blob/master/src/main/java/inra/ijpb/measure/region2d/MorphometricFeatures2D.java). It encapsulates an enum class [MorphometricFeatures2D.Feature](https://github.com/ijpb/MorphoLibJ/blob/master/src/main/java/inra/ijpb/measure/region2d/MorphometricFeatures2D.java#L62C17-L62C24). This enumeration lists the different features that can be computed with the AnalyzeRegions plugin. Then, the [computeTable method](https://github.com/ijpb/MorphoLibJ/blob/master/src/main/java/inra/ijpb/measure/region2d/MorphometricFeatures2D.java#L223C25-L223C37) populates the results table from an ImagePlus containing a 2D label map.
In practice, most  of the computations are performed within specialized feature analyzers (like [Centroid](https://github.com/ijpb/MorphoLibJ/blob/master/src/main/java/inra/ijpb/measure/region2d/Centroid.java), [EquivalentEllipse](https://github.com/ijpb/MorphoLibJ/blob/master/src/main/java/inra/ijpb/measure/region2d/EquivalentEllipse.java)...). The MorphometricFeatures2D.computeTable method calls them is the appropriate way (to manage dependencies), then populates the table from the various results.

**Comment by tischi** [comment]
I see!

```
    public MorphometricFeatures2D add(Feature f)
    {
        features.add(f);
        return this;
    }
```

I guess this does the trick?! Isn't that some kind of builder pattern?

**Comment by dlegland** [comment]
> I guess this does the trick?!

Yes! Works like classical methods from the Collections framework, but returns instance to "this".

>  Isn't that some kind of builder pattern?

To be honest, I do not really know... Returning an instance to this is typical from the Builder pattern, but the builder approaches I have seen often use two different classes: one for the builder, another one for the main class. Here, there is no "build()" method at the end.
Or maybe the MorphometricFeatures2D class can be seen as a builder for a ResultsTable!

**Comment by dlegland** [comment]
Hi!
a small follow-up, I have[ just pushed](4cd981ee88cfbba5ffa34cda0f69b0452a1fff94) the refactoring of the [AnalyzeRegions3D ](https://github.com/ijpb/MorphoLibJ/blob/master/src/main/java/inra/ijpb/plugins/AnalyzeRegions3D.java) plugin to use a new [MorphometricFeatures3D ](https://github.com/ijpb/MorphoLibJ/blob/master/src/main/java/inra/ijpb/measure/GeometricMeasures3D.java) based on the same principle that the 2D equivalent.

Everything seems to work fine: same behaviour as before, and it is now possible to setup features programmatically.

I am just not totally satisfied with management of options, but this will be for a future enhancement.

On the roadmap is also the refactoring of the dialogs for analysis of regions. This may take some time however.

**Comment by dlegland** [comment]
Hi there, 

another follow-up! I've just released a [new "RegionFeaturesJ" plugin](https://github.com/ijtools/RegionFeaturesJ), that aims at being moreflexible for the management of the features wecan compute on regions. In particular, it should be possible to add its own user-defined feature, and get a results table with all the feature results.

A typical syntax is as follow:
```java
ImagePlus image = createImagePlus();
ResultsTable table = RegionFeatures.initialize(image)
        .add(Area.class)
        .add(Perimeter.class)
        .add(Circularity.class)
        // potentially other features
        .createTable();
```
Still under development, but for 2D regions, it provides the same features as with MorphoLibJ (the RegionFeaturesJ plugin actually depends on MorphoLibJ). I expect it to be more flexible and exensible that the MorphoLibJ counterpart.

Feel free to try and to gove feedback!

**Comment by dlegland** [comment]
Hi,
I updated the [RegionFeaturesJ plugin](https://github.com/ijtools/RegionFeaturesJ), that now manages 3D morphological measurements, as well as intensity measurements. Both are largeley based on the MorphoLibJ code, but it is also an occasion to refresh them a bit...

They are also additional options:
* to include image name within results table (usefull for concatenating results)
* to include measurement unit within column names (to make sure how to interpret numerical values).

feedback welcome! (@tischi , @iarganda ...)

**Comment by tischi** [comment]
include image name: yes!

------

include measurement unit: Maybe better to spend an additional column on this?

For example:

| area | area_unit | 
|----------|----------|
| 101.4 | um2 | 

The advantage would be that downstream analysis scripts could rely on the column `area` always being there no matter what the unit is.

**Comment by dlegland** [comment]
Hi,
yes, I agree that the same feature would result in different column names, and that could be problematic in some cases. But on the other sides, measuring N features would results in a table with N*2 columns, and that will grow very quickly.
Maybe can be a multi-choice option?
Will think about this...

**Comment by tischi** [comment]
Another option would be an additional table, where the rows are the column names, and the columns are column metadata.

For example:

| column | unit | 
|----------|----------|
| area | um2 | 
| volume | um3 |
